lib/elixir/lib/task/supervised.ex

```diff
@@ -539,9 +531,55 @@ defmodule Task.Supervised do
 
       {:EXIT, _, _} ->
         stream_monitor_loop(running_tasks, config)
+
+      other ->
+        handle_stop_or_parent_down(other, running_tasks, config)
+        stream_monitor_loop(running_tasks, config)
     end
   end
 
+  defp stream_waiting_for_stop_loop(running_tasks, config) do
+    %{
+      parent_ref: parent_ref,
+      monitor_ref: monitor_ref
+    } = config
+
+    receive do
+      message ->
+        handle_stop_or_parent_down(message, running_tasks, config)
+        stream_waiting_for_stop_loop(running_tasks, config)
+    end
+  end
+
+  # The parent process is telling us to stop because the stream is being
+  # closed. In this case, we forcibly kill all spawned processes and then
+  # exit gracefully ourselves.
+  defp handle_stop_or_parent_down({:stop, monitor_ref}, running_tasks, %{
+         monitor_ref: monitor_ref
+       }) do
+    Process.flag(:trap_exit, true)
+
+    for {ref, %{pid: pid}} <- running_tasks do
+      Process.exit(pid, :kill)
+
+      receive do
+        {:DOWN, ^ref, _, _, _} -> :ok
+      end
+    end
+
+    exit(:normal)
+  end
+
+  defp handle_stop_or_parent_down({:DOWN, parent_ref, _, _, reason}, running_tasks, %{
+         parent_ref: parent_ref
+       }) do
```

Comment:
We rarely use this pattern in Elixir, can we please keep all arguments in a single line or one per line? Thanks. Same for the clauses above and then tests.